pkg/events/events.go

```diff
@@ -134,10 +142,30 @@ func ConsumeAll(s Consumer, group string) (<-chan Event, error) {
 
 // Publish publishes the ev to the MainQueue from where it is distributed to all subscribers
 // NOTE: needs to use reflect on runtime
-func Publish(s Publisher, ev interface{}) error {
+func Publish(ctx context.Context, s Publisher, ev interface{}) error {
 	evName := reflect.TypeOf(ev).String()
+	traceParent := getTraceParentFromCtx(ctx)
 	return s.Publish(MainQueueName, ev, events.WithMetadata(map[string]string{
-		MetadatakeyEventType: evName,
-		MetadatakeyEventID:   uuid.New().String(),
+		MetadatakeyEventType:   evName,
+		MetadatakeyEventID:     uuid.New().String(),
+		MetadatakeyTraceParent: traceParent,
 	}))
 }
+
+// GetTraceContext extracts the trace context from the event and injects it into the given
+// context.
+func (e *Event) GetTraceContext(ctx context.Context) context.Context {
+	return propagation.TraceContext{}.Extract(ctx, propagation.MapCarrier{
+		"traceparent": e.TraceParent,
+	})
+}
+
+// getTraceParentFromCtx will return a traceparent from the context if it exists.
+// it will be a string as specificied here: https://www.w3.org/TR/trace-context/
+// If no trace info in the context, the return will be an empty string
+func getTraceParentFromCtx(ctx context.Context) string {
+	mc := propagation.MapCarrier{}
+	tc := propagation.TraceContext{}
+	tc.Inject(ctx, &mc)
+	return mc["traceparent"]
```

Comment:
This should use the constant?

Reply:
The constant there is not the same as the other one, this is the key that is described in the wc3 spec and what is used by the propagation functions. I think using the same constant will give the wrong intent.

Reply:
Ah, ok.